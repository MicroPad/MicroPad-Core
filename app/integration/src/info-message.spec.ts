import { Page } from 'playwright';
import { openMicroPadRaw } from './utils/general-ui';

describe(`App Info Messages`, () => {
	it(`should not show the info banner if there is no informational message available`, async () => {
		// Arrange
		const page = await browser.newContext().then(ctx => ctx.newPage());
		await page.route(/info\.json\?.*$/, route => {
			route.fulfill({
				status: 404,
				headers: {
					'access-control-allow-origin': '*'
				}
			});
		});

		// Act
		await openMicroPad(page);
		await page.waitForResponse(/info\.json\?.*$/, { timeout: 6000 });

		// Assert
		expect(await page.$('.info-banner')).toBeNull();
	});

	it(`should not show the info banner if there is a malformed message`, async () => {
		// Arrange
		const page = await browser.newContext().then(ctx => ctx.newPage());
		await page.route(/info\.json\?.*$/, route => {
			route.fulfill({
				status: 200,
				contentType: 'application/json',
				headers: {
					'access-control-allow-origin': '*'
				},
				body: `this isn't valid JSON`
			});
		});

		// Act
		await openMicroPad(page);
		await page.waitForResponse(/info\.json\?.*$/, { timeout: 6000 });

		// Assert
		expect(await page.$('.info-banner')).toBeNull();
	});

	it(`should show the info banner if there is a message`, async () => {
		// Arrange
		const expected = 'Hello test software! beep boop';

		const page = await browser.newContext().then(ctx => ctx.newPage());
		await page.route(/info\.json\?.*$/, route => {
			route.fulfill({
				status: 200,
				contentType: 'application/json',
				headers: {
					'access-control-allow-origin': '*'
				},
				body: JSON.stringify({ text: expected })
			});
		});

		// Act
		await openMicroPad(page);
		await page.waitForResponse(/info\.json\?.*$/, { timeout: 6000 });
		const infoBanner = await page.$('.info-banner__msg');

		// Assert
		expect(await infoBanner?.innerText()).toBe(expected);
	});

	it(`should show the info banner with link if there is a message with a call-to-action`, async () => {
		// Arrange
		const expected = 'Hello test software! beep boop';
		const expectedCta = 'https://www.youtube.com/watch?v=dQw4w9WgXcQ';

		const page = await browser.newContext().then(ctx => ctx.newPage());
		await page.route(/info\.json\?.*$/, route => {
			route.fulfill({
				status: 200,
				contentType: 'application/json',
				headers: {
					'access-control-allow-origin': '*'
				},
				body: JSON.stringify({
					text: expected,
					cta: expectedCta
				})
			});
		});

		// Act
		await openMicroPad(page);
		await page.waitForResponse(/info\.json\?.*$/, { timeout: 6000 });
		const infoBanner = await page.$('.info-banner__msg');

		// Assert
		expect(await infoBanner?.innerHTML()).toMatchSnapshot();
	});
});

async function openMicroPad(page: Page) {
	await openMicroPadRaw(page);
	await page.waitForSelector('.brand-logo');
}
