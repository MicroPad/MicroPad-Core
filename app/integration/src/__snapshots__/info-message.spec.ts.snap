// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Info Messages should show the info banner with link if there is a message with a call-to-action 1`] = `"<strong>Hello test software! beep boop</strong><a class=\\"info-banner__cta\\" target=\\"_blank\\" rel=\\"noopener noreferrer nofollow\\" href=\\"https://www.youtube.com/watch?v=dQw4w9WgXcQ\\">Learn more.</a>"`;
